src/kinematics/DiffDriveKinematics.cpp

```diff
@@ -49,4 +51,34 @@ pose_t DiffDriveKinematics::getNextPose(const wheelvel_t& wheelVel, const pose_t
 										double dt) const {
 	return pose + getPoseUpdate(wheelVel, pose(2), dt);
 }
+
+pose_t DiffDriveKinematics::ensureWithinWheelSpeedLimit(
+	PreferredVelPreservation preferredVelPreservation, double xVel, double thetaVel,
+	double maxWheelSpeed) const {
+	auto wheelVels = robotVelToWheelVel(xVel, thetaVel);
+	double calculatedMaxWheelVel = std::max(wheelVels.lVel, wheelVels.rVel);
+	if (calculatedMaxWheelVel > maxWheelSpeed) {
+		double scaleFactor = maxWheelSpeed / calculatedMaxWheelVel;
+		switch (preferredVelPreservation) {
+			case PreferredVelPreservation::Proportional:
+				xVel *= scaleFactor;
+				thetaVel *= scaleFactor;
+				break;
+			case PreferredVelPreservation::PreferXVel:
+				thetaVel *= (0.5 * scaleFactor);
+				break;
+			case PreferredVelPreservation::PreferThetaVel:
+				xVel *= (0.5 * scaleFactor);
+				break;
```

Comment:
Don't think this math is right, since I don't think we can just do regular scaling here. Refer to the equation [here](https://github.com/huskyroboticsteam/Resurgence/pull/285/files#diff-b474567bf27f33d3c2f4b94d7af43bd3a081d63bb4017e7c2f0ca3b09aea37a3L16). 

If we're preferring xVel, then we should change thetaVel to satisfy the wheel speed constraint. Somewhat sure this pseudocode is correct.

```
if abs(xVel) > maxSpeed:
    return {copysign(maxSpeed, xVel), 0, 0}

if lVel < -maxSpeed:
    thetaVel = (-maxSpeed - xVel) * -2.0 / wheelBaseWidth
if lVel > maxSpeed:
    thetaVel = (maxSpeed - xVel) * -2.0 / wheelBaseWidth
if rVel < -maxSpeed:
    thetaVel = (-maxSpeed - xVel) * 2.0 / wheelBaseWidth
if rVel > maxSpeed:
    thetaVel = (maxSpeed - xVel) * 2.0 / wheelBaseWidth
```

And we do a similar thing for prefer theta vel, again somewhat sure about correctness:

```
if abs(wheelBaseWidth / 2 * thetaVel) > maxSpeed:
    return {0, 0, copysign(maxSpeed * 2 / wheelBaseWidth, thetaVel)}

if max(lVel, rVel) > maxSpeed:
    xVel -= max(lVel, rVel) - maxSpeed
if min(lVel, rVel) < -maxSpeed:
    xVel += -maxSpeed - min(lVel, rVel)
```